src/python/pants/backend/go/util_rules/analyze_package.go

```diff
@@ -0,0 +1,339 @@
+	TestImports  []string `json:",omitempty"`
+	XTestImports []string `json:",omitempty"`
+
+	// //go:embed patterns found in Go source files
+	// For example, if a source file says
+	//	//go:embed a* b.c
+	// then the list will contain those two strings as separate entries.
+	// (See package embed for more details about //go:embed.)
+	EmbedPatterns      []string `json:",omitempty"` // patterns from GoFiles, CgoFiles
+	TestEmbedPatterns  []string `json:",omitempty"` // patterns from TestGoFiles
+	XTestEmbedPatterns []string `json:",omitempty"` // patterns from XTestGoFiles
+
+	// Error information. This differs from how `go list` reports errors.
+	InvalidGoFiles map[string]string `json:",omitempty"`
+	Error          string            `json:",omitempty"`
+}
+
+type fileAnalysis struct {
+	Name    string
+	Imports []string
+}
+
+func analyzeFile(fileSet *token.FileSet, filename string) (*fileAnalysis, error) {
+	parsed, err := parser.ParseFile(fileSet, filename, nil, parser.ImportsOnly|parser.ParseComments)
+	if err != nil {
+		return nil, err
+	}
+
+	analysis := &fileAnalysis{}
+
+	analysis.Name = parsed.Name.Name
+	for _, spec := range parsed.Imports {
+		importPath, err := strconv.Unquote(spec.Path.Value)
+		if err != nil {
+			return nil, fmt.Errorf("unable to decode import %s in %s", spec.Path.Value, filename)
+		}
+		analysis.Imports = append(analysis.Imports, importPath)
+	}
+
+	return analysis, nil
+}
+
+// Copied from https://cs.opensource.google/go/go/+/refs/tags/go1.17.2:src/go/build/build.go;l=1024;drc=refs%2Ftags%2Fgo1.17.2
+func fileListForExt(p *Package, ext string) *[]string {
+	switch ext {
+	case ".c":
+		return &p.CFiles
+	case ".cc", ".cpp", ".cxx":
+		return &p.CXXFiles
+	case ".m":
+		return &p.MFiles
+	case ".h", ".hh", ".hpp", ".hxx":
+		return &p.HFiles
+	case ".f", ".F", ".for", ".f90":
+		return &p.FFiles
+	case ".s", ".S", ".sx":
+		return &p.SFiles
+	case ".swig":
+		return &p.SwigFiles
+	case ".swigcxx":
+		return &p.SwigCXXFiles
+	case ".syso":
+		return &p.SysoFiles
+	}
+	return nil
+}
+
+func cleanImports(importsMap map[string]bool) []string {
+	var imports []string
+	for importPath, _ := range importsMap {
+		imports = append(imports, importPath)
+	}
+	sort.Strings(imports)
+	return imports
+}
+
+func analyzePackage(directory string, buildContext *build.Context) (*Package, error) {
+	pkg := &Package{
+		InvalidGoFiles: make(map[string]string),
+	}
+
+	fileSet := token.NewFileSet()
+
+	entries, err := os.ReadDir(directory)
+	if err != nil {
+		return pkg, fmt.Errorf("failed to read directory %s: %s", directory, err)
+	}
+
+	// Keep track of the names used in `package` directives to ensure that only one package name is used.
+	packageNames := make(map[string]bool)
+
+	importsMap := make(map[string]bool)
+	testImportsMap := make(map[string]bool)
+	xtestImportsMap := make(map[string]bool)
+	var Sfiles []string // files with ".S"(capital S)/.sx(capital s equivalent for case insensitive filesystems)
```

Comment:
Maybe clarify these are for Cgo, like `CgoSFiles`? This confused me

Reply:
Renamed.